ccl/changefeedccl: TestChangefeedIdleness failed

ccl/changefeedccl.TestChangefeedIdleness [failed](https://tanzanite.cluster.engflow.com/invocations/default/d44c68cc-946c-456d-b2c2-cf5ac1b064a8?testReportRun=3&testReportShard=12&testReportAttempt=1#targets-Ly9wa2cvY2NsL2NoYW5nZWZlZWRjY2w6Y2hhbmdlZmVlZGNjbF90ZXN0) on release-24.1.7-rc @ [62d13cef08f5a5b487216dc9c996f01a866f1341](https://github.com/cockroachdb/cockroach/commits/62d13cef08f5a5b487216dc9c996f01a866f1341):


```
=== RUN   TestChangefeedIdleness
    test_log_scope.go:165: test logs captured to: outputs.zip/logTestChangefeedIdleness3570400521
    test_log_scope.go:76: use -show-logs to present logs inline
    changefeed_test.go:511: -- test log scope end --
test logs left over in: outputs.zip/logTestChangefeedIdleness3570400521
--- FAIL: TestChangefeedIdleness (236.07s)
=== RUN   TestChangefeedIdleness/webhook
    helpers_test.go:872: making server as system tenant
    helpers_test.go:948: making webhook feed factory
    changefeed_test.go:473: condition failed to evaluate within 3m45s: expected (0) idle changefeeds, found (2)
    --- FAIL: TestChangefeedIdleness/webhook (236.07s)
```

Parameters:
 - <code>attempt=1</code>
 - <code>race=true</code>
 - <code>run=3</code>
 - <code>shard=12</code>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/cdc
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestChangefeedIdleness.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-43721

--- a/pkg/ccl/changefeedccl/changefeed_test.go
+++ b/pkg/ccl/changefeedccl/changefeed_test.go
@@ -471,13 +471,33 @@ func TestChangefeedIdleness(t *testing.T) {
 
 		registry := s.Server.JobRegistry().(*jobs.Registry)
 		currentlyIdle := registry.MetricsStruct().JobMetrics[jobspb.TypeChangefeed].CurrentlyIdle
+		// Use a wait group for cases when the number of idle changefeeds temporarily
+		// decreases, to avoid a race condition where the changefeed becomes idle
+		// before the idleness is checked.
+		var wg sync.WaitGroup
 		waitForIdleCount := func(numIdle int64) {
+			wg.Add(1)
 			testutils.SucceedsSoon(t, func() error {
 				if currentlyIdle.Value() != numIdle {
 					return fmt.Errorf("expected (%+v) idle changefeeds, found (%+v)", numIdle, currentlyIdle.Value())
 				}
 				return nil
 			})
+			wg.Done()
+		}
+		done := make(chan bool)
+		workload := func() {
+			for {
+				select {
+				case <-done:
+					return
+				default:
+					sqlDB.Exec(t, `INSERT INTO foo VALUES (0)`)
+					sqlDB.Exec(t, `DELETE FROM foo WHERE a = 0`)
+					sqlDB.Exec(t, `INSERT INTO bar VALUES (0)`)
+					sqlDB.Exec(t, `DELETE FROM bar WHERE b = 0`)
+				}
+			}
 		}
 
 		sqlDB.Exec(t, `CREATE TABLE foo (a INT PRIMARY KEY)`)
@@ -486,9 +506,10 @@ func TestChangefeedIdleness(t *testing.T) {
 		cf2 := feed(t, f, "CREATE CHANGEFEED FOR TABLE bar WITH resolved='10ms'")
 		defer closeFeed(t, cf1)
 
-		sqlDB.Exec(t, `INSERT INTO foo VALUES (0)`)
-		sqlDB.Exec(t, `INSERT INTO bar VALUES (0)`)
-		waitForIdleCount(0)
+		go workload()
+		go waitForIdleCount(0)
+		wg.Wait()
+		done <- true
 		waitForIdleCount(2) // Both should eventually be considered idle
 
 		jobFeed := cf2.(cdctest.EnterpriseTestFeed)
@@ -501,14 +522,12 @@ func TestChangefeedIdleness(t *testing.T) {
 		closeFeed(t, cf2)
 		waitForIdleCount(1) // The cancelled changefeed isn't considered idle
 
-		sqlDB.Exec(t, `INSERT INTO foo VALUES (1)`)
-		waitForIdleCount(0)
+		go workload()
+		go waitForIdleCount(0)
+		wg.Wait()
+		done <- true
 		waitForIdleCount(1)
 
-		assertPayloads(t, cf1, []string{
-			`foo: [0]->{"after": {"a": 0}}`,
-			`foo: [1]->{"after": {"a": 1}}`,
-		})
 	}, feedTestEnterpriseSinks)
 }
 

